src/librustc_parse/parser/mod.rs

```diff
@@ -662,6 +662,21 @@ impl<'a> Parser<'a> {
                                 }
                             }
 
+                            // If this was a missing `@` in a binding pattern
+                            // bail with a suggestion
+                            // https://github.com/rust-lang/rust/issues/72373
+                            if self.prev_token.is_ident() && &self.token.kind == &token::DotDot {
+                                expect_err
+                                    .span_suggestion_short(
+                                        self.token.span,
+                                        &format!("maybe missing `@` in binding pattern?"),
```

Comment:
Ideally we would want to change the message to be something like "if you meant to bind the contents of the rest of the array pattern into `ts`, use `@`. It would also be beneficial to verify that we are in an array pattern (but if it makes the code much more convoluted, we can skip that check).

Reply:
0fdd31317a9c8d2fbc284546c711bc5d5139d83b

I was able to pull the name from the id in the previous token with `pprust::token_to_string`.

Reply:
> It would also be beneficial to verify that we are in an array pattern (but if it makes the code much more convoluted, we can skip that check).

Is this valid in tuples too?

Reply:
[No, only bare `..` are allowed (`(x, ..)`)](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=1fb02a7b485d2b58571112cb0286b97a).

Reply:
Hmm...  How would I go about checking that this is in an array only?

Reply:
It looks like that message could be cleaned up too:

```
error: `..` patterns are not allowed here
 --> src/main.rs:4:15
  |
4 |         (a, x@..) => {}
  |               ^^
  |
  = note: only allowed in tuple, tuple struct, and slice patterns
```

That note appears to be incorrect and the error should indicate `@ ..` are not allowed in the tuple, not `..`?

Reply:
I checked this for tuple structs and the same message appears.  I will open a new issue for this.

Reply:
https://github.com/rust-lang/rust/issues/72574